src/Twilio/Security/RequestValidator.cs

```diff
@@ -45,16 +45,65 @@ public bool Validate(string url, NameValueCollection parameters, string expected
         /// <returns>true if the signature matches the result; false otherwise</returns>
         public bool Validate(string url, IDictionary<string, string> parameters, string expected)
         {
-            // check signature of url with and without port, since sig generation on back end is inconsistent
-            var signatureWithoutPort = GetValidationSignature(RemovePort(url), parameters);
-            var signatureWithPort = GetValidationSignature(AddPort(url), parameters);
-            // If either url produces a valid signature, we accept the request as valid
-            return SecureCompare(signatureWithoutPort, expected) || SecureCompare(signatureWithPort, expected);
-        }
-
+            if (string.IsNullOrEmpty(url))
+                throw new ArgumentException("Parameter 'url' cannot be null or empty.");
+            if (string.IsNullOrEmpty(expected))
+                throw new ArgumentException("Parameter 'expected' cannot be null or empty.");
+
+            if(parameters == null || parameters.Count == 0)
+            {
+                var signature = GetValidationSignature(url);
+                if (SecureCompare(signature, expected)) return true;
+
+                // check signature of url with and without port, since sig generation on back end is inconsistent
+                // If either url produces a valid signature, we accept the request as valid
+                url = GetUriVariation(url);
+                signature = GetValidationSignature(url);
+                if (SecureCompare(signature, expected)) return true;
+                return false;
+            }
+            else
+            {
+                var parameterStringBuilder = GetJoinedParametersStringBuilder(parameters);
+                parameterStringBuilder.Insert(0, url);
+                var signature = GetValidationSignature(parameterStringBuilder.ToString());
+                if (SecureCompare(signature, expected)) return true;
+                parameterStringBuilder.Remove(0, url.Length);
+
+                // check signature of url with and without port, since sig generation on back end is inconsistent
+                // If either url produces a valid signature, we accept the request as valid
+                url = GetUriVariation(url);
+                parameterStringBuilder.Insert(0, url);
+                signature = GetValidationSignature(parameterStringBuilder.ToString());
+                if (SecureCompare(signature, expected)) return true;
+
+                return false;
+            }
+        }
+
+        private StringBuilder GetJoinedParametersStringBuilder(IDictionary<string, string> parameters)
+        {
+            var keys = new List<string>(parameters.Keys);
+            keys.Sort(StringComparer.Ordinal);

```

Comment:
```suggestion
            var keys = new string[parameters.Count];
            var index = 0;
            foreach (var pair in parameters)
            {
                keys[index++] = pair.Key;
            }
            Array.Sort(keys, StringComparer.Ordinal);

```

Reply:
I tried this but the benchmark showed it was slower, so I reverted to the original:
|              Method |      Mean |     Error |    StdDev |   Gen0 |   Gen1 | Allocated |
|-------------------- |----------:|----------:|----------:|-------:|-------:|----------:|
| OriginalUnhappyPath | 15.288 μs | 0.1718 μs | 0.1523 μs | 1.9836 | 0.2441 |  12.27 KB |
|  CurrentUnhappyPath | 12.631 μs | 0.1982 μs | 0.1854 μs | 1.4496 | 0.2289 |   8.91 KB |
|   OriginalHappyPath |  9.241 μs | 0.1789 μs | 0.2449 μs | 1.6327 | 0.2289 |  10.02 KB |
|    CurrentHappyPath |  4.183 μs | 0.0834 μs | 0.1703 μs | 0.6943 | 0.2289 |   4.28 KB |

However, using `parameters.Keys.ToArray()` and then doing the array sort is faster.
Unfortunately, `.ToArray()` isn't available on 3.5.

Reply:
> However, using parameters.Keys.ToArray() and then doing the array sort is faster.

HMMMM, what about:

```C#
var keys = new string[parameters.Count]
paramters.Keys.CopyTo(keys, 0);
Array.Sort(keys, StringComparer.Ordinal);
```

Reply:
Hmm, I got confused by something. `.ToArray()` does seem to be available on 3.5.
Your suggestion seems to have the same performance as `.ToArray()`.

|              Method |      Mean |     Error |    StdDev |   Gen0 |   Gen1 | Allocated |
|-------------------- |----------:|----------:|----------:|-------:|-------:|----------:|
| OriginalUnhappyPath | 15.439 us | 0.2463 us | 0.2304 us | 1.9836 | 0.2441 |  12.27 KB |
|  CurrentUnhappyPath | 12.172 us | 0.2259 us | 0.2113 us | 1.4496 | 0.2289 |   8.88 KB |
|   OriginalHappyPath |  9.110 us | 0.1754 us | 0.1641 us | 1.6327 | 0.2289 |  10.02 KB |
|    CurrentHappyPath |  3.948 us | 0.0786 us | 0.1868 us | 0.6866 | 0.2213 |   4.23 KB |

Since `.ToArray()` is available, I'll go with that.